### PR TITLE
KAFKA-16682: Rewrite JaasTestUtils by Java

### PR DESCRIPTION
**Pull request by frankvicky**
To enable ClusterTestExtensions supporting SASL and leverage new test infra, we should rewrite `JassTestUtils` by Java

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+        }
+        return new ScramLoginModule(scramUser, scramPassword, false, new HashMap<>()).toString();
+    }
+
+    public static String clientLoginModule(String mechanism, Optional<File> keytabLocation, String serviceName) {
+        return kafkaClientModule(mechanism, keytabLocation, KAFKA_CLIENT_PRINCIPAL, KAFKA_PLAIN_USER, KAFKA_PLAIN_PASSWORD, KAFKA_SCRAM_USER, KAFKA_SCRAM_PASSWORD, KAFKA_OAUTH_BEARER_USER, serviceName).toString();
+    }
+
+    public static String clientLoginModule(String mechanism, Optional<File> keytabLocation) {
+        return clientLoginModule(mechanism, keytabLocation, SERVICE_NAME).toString();
```
Comment (chia7712) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: `toString` is redundant.

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+    }
+
+    public static String adminLoginModule(String mechanism, Optional<File> keytabLocation, String serviceName) {
+        return kafkaClientModule(mechanism, keytabLocation, KAFKA_SERVER_PRINCIPAL, KAFKA_PLAIN_ADMIN, KAFKA_PLAIN_ADMIN_PASSWORD, KAFKA_SCRAM_ADMIN, KAFKA_SCRAM_ADMIN_PASSWORD, KAFKA_OAUTH_BEARER_ADMIN, serviceName).toString();
+    }
+
+    public static String adminLoginModule(String mechanism, Optional<File> keytabLocation) {
+        return adminLoginModule(mechanism, keytabLocation, SERVICE_NAME).toString();
```
Comment (chia7712) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: `toString` is redundant.

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+                if (ScramMechanism.fromMechanismName(mechanism) != ScramMechanism.UNKNOWN) {
+                    return new ScramLoginModule(scramUser, scramPassword, false, new HashMap<>());
+                } else {
+                    throw new IllegalArgumentException("Unsupported client mechanism " + mechanism);
+                }
+        }
+    }
+
+    public static JaasSection kafkaClientSection(Optional<String> mechanism, Optional<File> keytabLocation) {
+        return new JaasSection(KAFKA_CLIENT_CONTEXT_NAME, mechanism.map(m -> kafkaClientModule(m, keytabLocation, KAFKA_CLIENT_PRINCIPAL_2, KAFKA_PLAIN_USER_2, KAFKA_PLAIN_PASSWORD_2, KAFKA_SCRAM_USER_2, KAFKA_SCRAM_PASSWORD_2, KAFKA_OAUTH_BEARER_USER_2, SERVICE_NAME)).map(Collections::singletonList).orElse(Collections.emptyList()));
```
Comment (chia7712) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: this line is too long ...

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+                return new Krb5LoginModule(
+                        true,
+                        true,
+                        keytabLocation.orElseThrow(() -> new IllegalArgumentException("Keytab location not specified for GSSAPI")).getAbsolutePath(),
+                        clientPrincipal,
+                        true,
+                        Optional.of(serviceName),
+                        IS_IBM_SECURITY
+                );
+            case "PLAIN":
```
Comment (chia7712) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: Could you please use `PlainSaslServer.PLAIN_MECHANISM` instead? and other constants should be replaced by existent variable too.

File: core/src/test/scala/unit/kafka/security/authorizer/AclAuthorizerWithZkSaslTest.scala [hunk 5]
```diff
@@ -180,7 +180,7 @@ class TestableDigestLoginModule extends DigestLoginModule {
   }
 }
 
-class TestableJaasSection(contextName: String, modules: Seq[JaasModule]) extends JaasSection(contextName, modules) {
+class TestableJaasSection(contextName: String, modules: Seq[JaasModule]) extends JaasSection(contextName, modules.asJava) {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/security/authorizer/AclAuthorizerWithZkSaslTest.scala]: `modules` can use java type to avoid extra conversion

File: core/src/test/scala/integration/kafka/api/CustomQuotaCallbackTest.scala [hunk 6]
```diff
@@ -46,31 +47,31 @@ class CustomQuotaCallbackTest extends IntegrationTestHarness with SaslSetup {
...
 
-  override protected lazy val trustStoreFile = Some(TestUtils.tempFile("truststore", ".jks"))
+  override protected lazy val trustStoreFile: Some[File] = Some(TestUtils.tempFile("truststore", ".jks"))
   override val brokerCount: Int = 2
 
   private val kafkaServerSaslMechanisms = Seq("SCRAM-SHA-256")
   private val kafkaClientSaslMechanism = "SCRAM-SHA-256"
-  override protected val serverSaslProperties = Some(kafkaServerSaslProperties(kafkaServerSaslMechanisms, kafkaClientSaslMechanism))
-  override protected val clientSaslProperties = Some(kafkaClientSaslProperties(kafkaClientSaslMechanism))
+  override protected val serverSaslProperties: Some[Properties] = Some(kafkaServerSaslProperties(kafkaServerSaslMechanisms, kafkaClientSaslMechanism))
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/CustomQuotaCallbackTest.scala]: please remove unnecessary changes

File: core/src/test/scala/integration/kafka/api/CustomQuotaCallbackTest.scala [hunk 7]
```diff
@@ -362,8 +364,8 @@ object GroupedUserQuotaCallback {
 class GroupedUserQuotaCallback extends ClientQuotaCallback with Reconfigurable with Logging {
 
   var brokerId: Int = -1
-  val customQuotasUpdated = ClientQuotaType.values.map(quotaType => quotaType -> new AtomicBoolean).toMap
-  val quotas = ClientQuotaType.values.map(quotaType => quotaType -> new ConcurrentHashMap[String, Double]).toMap
+  val customQuotasUpdated: Map[ClientQuotaType, AtomicBoolean] = ClientQuotaType.values.map(quotaType => quotaType -> new AtomicBoolean).toMap
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/CustomQuotaCallbackTest.scala]: ditto

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for this nice patch!

**Comment by frankvicky** [comment]
Hello @chia7712 
I have addressed all comment, PTAL 😺

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+import java.util.Optional;
+import java.util.Properties;
+
+import static org.apache.kafka.common.config.SaslConfigs.GSSAPI_MECHANISM;
+import static org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule.OAUTHBEARER_MECHANISM;
+import static org.apache.kafka.common.security.plain.internals.PlainSaslServer.PLAIN_MECHANISM;
+
+public class JaasTestUtils {
+
+    public static class Krb5LoginModule implements JaasModule {
```
Comment (chia7712) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: Could you please move those impls of `JaasModule` to separate class? for example:
```java
public class JaasModule {
    public static JaasModule zkDigestModule(boolean debug, Map<String, String> entries) {
        String name = "org.apache.zookeeper.server.auth.DigestLoginModule";
        return new JaasModule(
            name,
            debug,
            entries,
            String.format("%s required\n  debug=%b\n  %s;\n", name, debug, entries.entrySet().stream()
                .map(e -> e.getKey() + "=\"" + e.getValue() + "\"")
                .reduce((e1, e2) -> e1 + "\n  " + e2)
                .orElse(""))
        );
    }

    private final String name;

    private final boolean debug;

    private final Map<String, String> entries;
    
    private final String toString;

    private JaasModule(String name, boolean debug, Map<String, String> entries, String toString) {
        this.name = name;
        this.debug = debug;
        this.entries = entries;
        this.toString = toString;
    }

    String name() {
        return name;
    }

    boolean debug() {
        return debug;
    }

    Map<String, String> entries() {
        return entries;
    }
    
    @Override
    public String toString() {
        return toString;
    }
}
```
It seems to me that is more readable. Also, we don't need to have many class declaration.

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for updated patch. I leave a major comment for this patch. PTAL

**Comment by frankvicky** [comment]
Hi @chia7712 
I have extract the `JaasModule` and do some refactors, PTAL 🐧

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+                        KAFKA_PLAIN_USER_2,
+                        KAFKA_PLAIN_PASSWORD_2,
+                        KAFKA_SCRAM_USER_2,
+                        KAFKA_SCRAM_PASSWORD_2,
+                        KAFKA_OAUTH_BEARER_USER_2,
+                        SERVICE_NAME)
+                ).map(Collections::singletonList).orElse(Collections.emptyList()));
+    }
+
+    private static String jaasSectionsToString(List<JaasSection> jaasSections) {
```
Comment (chia7712) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: Could you please merge those two methods?
```java
    private static void writeToFile(File file, List<JaasSection> jaasSections) throws IOException {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
            writer.write(String.join("", jaasSections.stream().map(Object::toString).toArray(String[]::new)));
        }
    }
```

File: core/src/test/scala/integration/kafka/api/SaslMultiMechanismConsumerTest.scala [hunk 10]
```diff
@@ -27,14 +30,14 @@ class SaslMultiMechanismConsumerTest extends BaseConsumerTest with SaslSetup {
   private val kafkaServerSaslMechanisms = List("GSSAPI", "PLAIN")
   this.serverConfig.setProperty(ZkConfigs.ZK_ENABLE_SECURE_ACLS_CONFIG, "true")
   override protected def securityProtocol = SecurityProtocol.SASL_SSL
-  override protected lazy val trustStoreFile = Some(TestUtils.tempFile("truststore", ".jks"))
-  override protected val serverSaslProperties = Some(kafkaServerSaslProperties(kafkaServerSaslMechanisms, kafkaClientSaslMechanism))
-  override protected val clientSaslProperties = Some(kafkaClientSaslProperties(kafkaClientSaslMechanism))
+  override protected lazy val trustStoreFile: Some[File] = Some(TestUtils.tempFile("truststore", ".jks"))
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslMultiMechanismConsumerTest.scala]: Please revert those unconnected changes.

File: core/src/test/scala/integration/kafka/api/SaslGssapiSslEndToEndAuthorizationTest.scala [hunk 11]
```diff
@@ -16,23 +16,21 @@
...
   override val clientPrincipal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE,
-    JaasTestUtils.KafkaClientPrincipalUnqualifiedName)
+    JaasTestUtils.KAFKA_CLIENT_PRINCIPAL_UNQUALIFIED_NAME)
   override val kafkaPrincipal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE,
-    JaasTestUtils.KafkaServerPrincipalUnqualifiedName)
+    JaasTestUtils.KAFKA_SERVER_PRINCIPAL_UNQUALIFIED_NAME)
 
   override protected def kafkaClientSaslMechanism = "GSSAPI"
-  override protected def kafkaServerSaslMechanisms = List("GSSAPI")
+  override protected def kafkaServerSaslMechanisms: List[String] = List("GSSAPI")
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslGssapiSslEndToEndAuthorizationTest.scala]: ditto

File: core/src/test/scala/integration/kafka/api/SaslPlainSslEndToEndAuthorizationTest.scala [hunk 12]
```diff
@@ -113,15 +115,15 @@ class SaslPlainSslEndToEndAuthorizationTest extends SaslEndToEndAuthorizationTes
...
+  val mechanismPrefix: String = listenerName.saslMechanismConfigPrefix("PLAIN")
   this.serverConfig.put(s"$mechanismPrefix${BrokerSecurityConfigs.SASL_SERVER_CALLBACK_HANDLER_CLASS_CONFIG}", classOf[TestServerCallbackHandler].getName)
   this.producerConfig.put(SaslConfigs.SASL_CLIENT_CALLBACK_HANDLER_CLASS, classOf[TestClientCallbackHandler].getName)
   this.consumerConfig.put(SaslConfigs.SASL_CLIENT_CALLBACK_HANDLER_CLASS, classOf[TestClientCallbackHandler].getName)
   this.adminClientConfig.put(SaslConfigs.SASL_CLIENT_CALLBACK_HANDLER_CLASS, classOf[TestClientCallbackHandler].getName)
   this.superuserClientConfig.put(SaslConfigs.SASL_CLIENT_CALLBACK_HANDLER_CLASS, classOf[TestClientCallbackHandler].getName)
 
   override protected def kafkaClientSaslMechanism = "PLAIN"
-  override protected def kafkaServerSaslMechanisms = List("PLAIN")
+  override protected def kafkaServerSaslMechanisms: List[String] = List("PLAIN")
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslPlainSslEndToEndAuthorizationTest.scala]: ditto

File: core/src/test/scala/integration/kafka/api/SaslSetup.scala [hunk 13]
```diff
@@ -62,18 +66,28 @@ trait SaslSetup {
   def startSasl(jaasSections: Seq[JaasSection]): Unit = {
     // Important if tests leak consumers, producers or brokers
     LoginManager.closeAll()
-    val hasKerberos = jaasSections.exists(_.modules.exists {
-      case _: Krb5LoginModule => true
-      case _ => false
-    })
+
+    val hasKerberos = jaasSections.exists { section =>
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSetup.scala]:
```scala
val hasKerberos = jaasSections.exists(_.getModules.asScala.exists(_.name().endsWith("Krb5LoginModule")))
```

File: core/src/test/scala/integration/kafka/api/SaslSetup.scala [hunk 14]
```diff
@@ -62,18 +66,28 @@ trait SaslSetup {
...
       initializeKerberos()
     }
+
     writeJaasConfigurationToFile(jaasSections)
-    val hasZk = jaasSections.exists(_.modules.exists {
-      case _: ZkDigestModule => true
-      case _ => false
-    })
+
+    val hasZk = jaasSections.exists { section =>
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSetup.scala]:
```scala
val hasZk = jaasSections.exists(_.getModules.asScala.exists(_.name() == "org.apache.zookeeper.server.auth.DigestLoginModule"))
```

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for updated patch

**Comment by frankvicky** [comment]
Hi @chia7712 
I have reverted those unnecessary changes, PTAL 😺

File: core/src/test/java/kafka/security/JaasModule.java [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+    public String name() {
+        return name;
+    }
+
+    public boolean debug() {
+        return debug;
+    }
+
+    Map<String, String> entries() {
```
Comment (OmniaGM) [inline core/src/test/java/kafka/security/JaasModule.java]: Is this method really used?

File: core/src/test/java/kafka/security/JaasModule.java [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+
+    private final String name;
+
+    private final boolean debug;
+
+    private final Map<String, String> entries;
+
+    private final String toString;
+
+    private JaasModule(String name, boolean debug, Map<String, String> entries, String toString) {
```
Comment (OmniaGM) [inline core/src/test/java/kafka/security/JaasModule.java]: Why are we sending `toString` to the class and not generating this in `toString` method? specially that as far as I can see all `toString` passed values has the same formate.

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+
+    public static String scramClientLoginModule(String mechanism, String scramUser, String scramPassword) {
+        if (ScramMechanism.fromMechanismName(mechanism) == ScramMechanism.UNKNOWN) {
+            throw new IllegalArgumentException("Unsupported SCRAM mechanism " + mechanism);
+        }
+        return JaasModule.scramLoginModule(scramUser, scramPassword, false, new HashMap<>()).toString();
+    }
+
+    public static String clientLoginModule(String mechanism, Optional<File> keytabLocation, String serviceName) {
+        return kafkaClientModule(mechanism, keytabLocation, KAFKA_CLIENT_PRINCIPAL, KAFKA_PLAIN_USER, KAFKA_PLAIN_PASSWORD, KAFKA_SCRAM_USER, KAFKA_SCRAM_PASSWORD, KAFKA_OAUTH_BEARER_USER, serviceName).toString();
```
Comment (OmniaGM) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: This line is very long can you please break it into multiple lines

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    public static String clientLoginModule(String mechanism, Optional<File> keytabLocation) {
+        return clientLoginModule(mechanism, keytabLocation, SERVICE_NAME);
+    }
+
+    public static String adminLoginModule(String mechanism, Optional<File> keytabLocation, String serviceName) {
+        return kafkaClientModule(mechanism, keytabLocation, KAFKA_SERVER_PRINCIPAL, KAFKA_PLAIN_ADMIN, KAFKA_PLAIN_ADMIN_PASSWORD, KAFKA_SCRAM_ADMIN, KAFKA_SCRAM_ADMIN_PASSWORD, KAFKA_OAUTH_BEARER_ADMIN, serviceName).toString();
```
Comment (OmniaGM) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: same

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+                    } else {
+                        throw new IllegalArgumentException("Unsupported server mechanism " + mechanism);
+                    }
+                    break;
+            }
+        }
+        return new JaasSection(contextName, modules);
+    }
+
+    private static JaasModule kafkaClientModule(String mechanism, Optional<File> keytabLocation, String clientPrincipal, String plainUser, String plainPassword, String scramUser, String scramPassword, String oauthBearerUser, String serviceName) {
```
Comment (OmniaGM) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: Can you also break the list of these attributes to multiple lines

**Review by OmniaGM — COMMENTED** [review]
Thanks for the PR I had a quick look and left few nits

File: core/src/test/java/kafka/security/JaasModule.java (hunk 16 above)
Comment (frankvicky) [reply core/src/test/java/kafka/security/JaasModule.java]: Thanks for the suggestion, I have remove those redundant `toString` and have only one in constructor. 😺

**Comment by frankvicky** [comment]
Hello @OmniaGM 
Thanks for the review, I have done some refactoring based on feedback, PTAL 😺

File: core/src/test/java/kafka/security/JaasTestUtils.java [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+
+        public String getContextName() {
+            return contextName;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("%s {\n  %s\n};\n",
+                    contextName,
+                    String.join("\n  ", modules.stream().map(Object::toString).toArray(String[]::new)));
```
Comment (chia7712) [inline core/src/test/java/kafka/security/JaasTestUtils.java]: `modules.stream().map(Object::toString).collect(Collectors.joining("\n  "))`

File: core/src/test/java/kafka/security/JaasModule.java [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+    private final String name;
+
+    private final boolean debug;
+
+    private final String toString;
+
+    private JaasModule(String name, boolean debug, Map<String, String> entries) {
+        this.name = name;
+        this.debug = debug;
+        this.toString = String.format("%s required\n  debug=%b\n  %s;\n", name, debug, entries.entrySet().stream()
```
Comment (chia7712) [inline core/src/test/java/kafka/security/JaasModule.java]:
```java
        this.toString = String.format("%s required\n  debug=%b\n  %s;\n", name, debug, entries.entrySet().stream()
                .map(e -> e.getKey() + "=\"" + e.getValue() + "\"")
                .collect(Collectors.joining("\n  ")));
```

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for updates. two small comments are left. PTAL

File: core/src/test/java/kafka/security/JaasTestUtils.java (hunk 20 above)
Comment (frankvicky) [reply core/src/test/java/kafka/security/JaasTestUtils.java]: Hi @chia7712 
I have test the output in my local machine, it seems that the suggestions would have different output compare to original one.

Before:
```
KafkaServer {
  org.apache.kafka.common.security.plain.PlainLoginModule required
  debug=false
  password=""
  username="plain-admin";

};
```

After:
```
org.apache.kafka.common.security.plain.PlainLoginModule required
  debug=false
  password=""
  username="plain-admin";
```

File: core/src/test/java/kafka/security/JaasModule.java (hunk 16 above)
Comment (chia7712) [reply core/src/test/java/kafka/security/JaasModule.java]: @frankvicky It seems you don't address @OmniaGM comment. 
```java
    @Override
    public String toString() {
        return String.format("%s required\n  debug=%b\n  %s;\n", name, debug, entries.entrySet().stream()
                .map(e -> e.getKey() + "=\"" + e.getValue() + "\"")
                .reduce((e1, e2) -> e1 + "\n  " + e2)
                .orElse(""));;
    }
```

Comment (frankvicky) [reply core/src/test/java/kafka/security/JaasModule.java]: Hi @chia7712 
It seems that I misunderstand the comment.
I utilize the `toString` field and build the `toString` value in constructor but it seems silly now. 
I will move the impl to `toString` method.

**Comment by frankvicky** [comment]
Hi @chia7712 
I have addressed the comment, PTAL 😸

File: core/src/test/java/kafka/security/JaasTestUtils.java (hunk 20 above)
Comment (chia7712) [reply core/src/test/java/kafka/security/JaasTestUtils.java]: Pardon me, could you share the test code to me? I don't notice the format like "KafkaServer { .. " on my local

Comment (frankvicky) [reply core/src/test/java/kafka/security/JaasTestUtils.java]: The tests which extend `EndToEndAuthorizationTest` will fail due to error config, it seems that these tests utilize the `toString` method to generate configuration

Comment (chia7712) [reply core/src/test/java/kafka/security/JaasTestUtils.java]: > The tests which extend EndToEndAuthorizationTest will fail due to error config, 

which one? I run a part of tests, and they pass with following changes.

```java
        @Override
        public String toString() {
            return String.format("%s {\n  %s\n};\n",
                    contextName,
                    modules.stream().map(Object::toString).collect(Collectors.joining("\n  ")));
        }
```

Comment (frankvicky) [reply core/src/test/java/kafka/security/JaasTestUtils.java]: I have misunderstanding of the comment, it works now. Sorry for that 😿

**Review by chia7712 — APPROVED** [review]
LGTM
